api/src/backend/tasks/tasks.py

```diff
@@ -251,84 +256,110 @@ def generate_outputs(scan_id: str, provider_id: str, tenant_id: str):
-
-    # Process findings in batches
-    for batch, is_last_batch in batched(
-        findings_qs.iterator(), DJANGO_FINDINGS_BATCH_SIZE
-    ):
-        finding_outputs = [
-            FindingOutput.transform_api_finding(finding, prowler_provider)
-            for finding in batch
-        ]
-
-        # Generate output files
-        for mode, config in OUTPUT_FORMATS_MAPPING.items():
-            kwargs = dict(config.get("kwargs", {}))
+    qs = Finding.all_objects.filter(scan_id=scan_id).order_by("uid").iterator()
+    for batch, is_last in batched(qs, DJANGO_FINDINGS_BATCH_SIZE):
+        fos = [FindingOutput.transform_api_finding(f, prowler_provider) for f in batch]
+
+        # Outputs
+        for mode, cfg in OUTPUT_FORMATS_MAPPING.items():
+            cls = cfg["class"]
+            suffix = cfg["suffix"]
+            extra = cfg.get("kwargs", {}).copy()
             if mode == "html":
-                kwargs["provider"] = prowler_provider
-                kwargs["stats"] = scan_summary
-
-            writer_class = config["class"]
-            if writer_class in output_writers:
-                writer = output_writers[writer_class]
-                writer.transform(finding_outputs)
-                writer.close_file = is_last_batch
-            else:
-                writer = writer_class(
-                    findings=finding_outputs,
-                    file_path=output_directory,
-                    file_extension=config["suffix"],
+                extra.update(provider=prowler_provider, stats=scan_summary)
+
+            writer, initialization = get_writer(
+                output_writers,
+                cls,
+                lambda cls=cls, fos=fos, suffix=suffix: cls(
+                    findings=fos,
+                    file_path=out_dir,
+                    file_extension=suffix,
                     from_cli=False,
-                )
-                writer.close_file = is_last_batch
-                output_writers[writer_class] = writer
-
-            # Write the current batch using the writer
-            writer.batch_write_data_to_file(**kwargs)
-
-            # TODO: Refactor the output classes to avoid this manual reset
-            writer._data = []
-
-    # Compress output files
-    output_directory = _compress_output_files(output_directory)
+                ),
+                is_last,
+            )
+            if not initialization:
+                writer.transform(fos)
+            writer.batch_write_data_to_file(**extra)
+            writer._data.clear()
+
+        # Compliance CSVs
+        for name in frameworks_avail:
+            compliance_obj = frameworks_bulk[name]
+            klass = next(
+                (
+                    c
+                    for cond, c in COMPLIANCE_CLASS_MAP.get(provider_type, [])
+                    if cond(name)
+                ),
+                GenericCompliance,
+            )
+            filename = f"{comp_dir}_{name}.csv"
```

Comment:
Again, to me this is really hard to follow.

Reply:
In order to simplify this implies more code and more testing, the MAP with the compliance is a dictionary whose key is the provider_type and inside a list of tuples, where the first element is the condition and the second one is the class

Then what we do here is to iterate that list and check the condition, if the condition is True we use the class, if none is True we use the Generic. I think that the formatting in this case of ruff makes it difficult to read

Reply:
If you want we can change to this version, it's the same:

```
for condition, cls in COMPLIANCE_CLASS_MAP.get(provider_type, []):
        if condition(framework_name):
            return cls
    return GenericCompliance
```

Reply:
To me @vicferpoy has the final word on this PR.

Reply:
I believe Adri already changed this. I agree it's better this way. We're getting older and I would have had issues trying to understand that piece of code in a few months 🤣

Reply:
Totally agree, I'm getting older.